### ISSUE TITLE
Web Api validation with custom validators using Unity DI not using custom validator - bug or feature?

### ISSUE THREAD

**Issue by truthbeliever** [JeremySkinner/FluentValidation#131]
Hi there

I think I found a bug.

I have some model
[Validator(typeof(AccompanyingDocumentModelValidator))]
	public class AccompanyingDocumentModel 
	{

and the Validator
public class AccompanyingDocumentModelValidator : AbstractValidator<AccompanyingDocumentModel>
	{

registered with Unity in OWIN Startup
FluentValidationModelValidatorProvider
				.Configure(config, provider => provider.ValidatorFactory = container.Resolve<IValidatorFactory>());

The VAlidatorFactory:
public class UnityValidatorFactory : IValidatorFactory
	{
		private readonly IUnityContainer _container;

		public UnityValidatorFactory()
		{
			_container = null;
		}

		public UnityValidatorFactory(IUnityContainer container)
		{
			_container = container;
		}

		public IValidator CreateInstance(Type validatorType)
		{
			return _container.Resolve(validatorType) as IValidator;
		}

		public IValidator<T> GetValidator<T>()
		{
			return (IValidator<T>)this.GetValidator(typeof(T));
		}
		
		/// <summary>
		/// Gets a validator for the appropriate type.
		/// </summary>
		public IValidator GetValidator(Type type)
		{
			if (type == null)
				return null;

			var attribute = (ValidatorAttribute)Attribute.GetCustomAttribute(type, typeof(ValidatorAttribute));

			if (attribute == null || attribute.ValidatorType == null)
				return this.CreateInstance(typeof(IValidator<>).MakeGenericType(type));

			return this.CreateInstance(attribute.ValidatorType) as IValidator;
		}

	}

and UnitiyContainer filled:	
p_Container.RegisterType<IValidator<AccompanyingDocumentModel>, AccompanyingDocumentModelValidator>(new HierarchicalLifetimeManager());
p_Container.RegisterType IValidatorFactory, 
UnityValidatorFactory>(new HierarchicalLifetimeManager());

This results in FluentValidationBodyModelValidator calls
private bool ValidateNodeAndChildren(ModelMetadata metadata, ValidationContext validationContext, object container)
		{
			object model = metadata.Model;
			bool isValid = true;

			// Optimization: we don't need to recursively traverse the graph for null and primitive types
			if (model == null || IsSimpleType(model.GetType()))
			{
				return ShallowValidate(metadata, validationContext, container);
			}

			// Check to avoid infinite recursion. This can happen with cycles in an object graph.
			if (validationContext.Visited.Contains(model))
			{
				return true;
			}
			validationContext.Visited.Add(model);

			// Validate the children first - depth-first traversal
			IEnumerable enumerableModel = model as IEnumerable;
			if (enumerableModel == null)
			{
				isValid = this.ValidateProperties(metadata, validationContext);
			}
			else
			{
				isValid = this.ValidateElements(enumerableModel, validationContext);
			}

			// Validate this node as well
			isValid = ShallowValidate(metadata, validationContext, container) && isValid;

			// Pop the object so that it can be validated again in a different path
			validationContext.Visited.Remove(model);

			return isValid;
		}

The bug is that it does not use the custom  validator but some standard ones( IValidator<string>) for every single Property of the APiControllers Method parameter.

So I changed it to the following checking for metadata's model Custom validation attribute. if present, it goes straight to the validation of the VAlidator decorated model using the assigned custom validator.
private bool ValidateNodeAndChildren(ModelMetadata metadata, ValidationContext validationContext, object container)
		{
			object model = metadata.Model;
			bool isValid = true;

			// Optimization: we don't need to recursively traverse the graph for null and primitive types
			if (model == null || IsSimpleType(model.GetType()))
			{
				return ShallowValidate(metadata, validationContext, container);
			}

			// Check to avoid infinite recursion. This can happen with cycles in an object graph.
			if (validationContext.Visited.Contains(model))
			{
				return true;
			}
			validationContext.Visited.Add(model);
			
			if (metadata.ModelType.GetCustomAttribute(typeof(ValidatorAttribute), false) != null)
			{
				isValid = ShallowValidate(metadata, validationContext, container) && isValid;
			}
			else
			{
				// Validate the children first - depth-first traversal
				IEnumerable enumerableModel = model as IEnumerable;
				if (enumerableModel == null)
				{
					isValid = this.ValidateProperties(metadata, validationContext);
				}
				else
				{
					isValid = this.ValidateElements(enumerableModel, validationContext);
				}

				// Validate this node as well
				isValid = ShallowValidate(metadata, validationContext, container) && isValid;
			}

			// Pop the object so that it can be validated again in a different path
			validationContext.Visited.Remove(model);

			return isValid;
		}

Did I miss something? or is this indeed a bug?

**Comment by JeremySkinner** [comment]
I'm a bit confused with what you're trying to do.

The ValidatorAttribute is for use with the AttributedValidatorFactory. When you ask it to create a validator, it will check to see if that attribute is present on the type and create the corresponding validator.

If you're using an IoC container, then the ValidatorAttribute is redundant. The entire point of using the container is that it is responsible for assocating validators with their type - you shouldn't be using the attribute for this. 

Basically, if you want to associate a specific validator with a type, you should be registering this with your container (I don't use unity, so I can't help with the exact code for this), not using the ValidatorAttribute.

**Comment by davidmrichter** [comment]
I registerd the validator like
p_Container.RegisterType IValidator<AccompanyingDocumentModel>, AccompanyingDocumentModelValidator (new HierarchicalLifetimeManager());

Then 
public class AccompanyingDocumentModel : IValidatableObject
{
...
public AccompanyingDocumentModel(IValidator<AccompanyingDocumentModel> p_Validator)
		{
			this.m_Validator = p_Validator;
		}

		public IEnumerable<ValidationResult> Validate(ValidationContext p_ValidationContext)
		{
			return this.m_Validator.Validate(this).ToValidationResult();
		}

But at the time the constructor got called the validator was not yet resolved, being null.

What did I wrong? How did you registering the validator with your IOC Framework? Even if it is not unity I could "translate" it to unity.

**Comment by davidmrichter** [comment]
And how does the FluentValidationBodyModelValidator Validate Method know that there is validator registered for that particular type not digging deeper using ValidateNodeAndChildren?

**Comment by JeremySkinner** [comment]
It calls into the ModelValidatorProvider, which in turn uses the ValidatorFactory that you specify in your call to FluentValidationModelValidatorProvider.Configure to ask for the correct validator.

With StructureMap I'd have a registry that calls ``For<IValidator<MyType>>().Use<MyValidator>()``

**Comment by davidmrichter** [comment]
I think i did the same with unity
`
p_Container.RegisterType<IValidatorFactory, UnityValidatorFactory>(new HierarchicalLifetimeManager());
			p_Container.RegisterType<IValidator<AccompanyingDocumentModel>, AccompanyingDocumentModelValidator>(new HierarchicalLifetimeManager());`

**Comment by JeremySkinner** [comment]
OK, so if you trace through the code, is your UnityValidatorFactory's GetValidator method being invoked?

**Comment by davidmrichter** [comment]
yes but only with string as requested type

**Comment by JeremySkinner** [comment]
OK that doesn't sound right - it should be called with your model type. What does your controller action look like?

**Comment by davidmrichter** [comment]
`[ValidationResponseFilter]		
		public async Task<IHttpActionResult> Post(AccompanyingDocumentModel p_AccompanyingDocumentModel)
`

**Comment by JeremySkinner** [comment]
Hard to tell what the problem could be. If you could put together a (small) sample project that reproduces the issue then I can look into it further.

**Comment by davidmrichter** [comment]
How do you register ein WEb Api OWIN? Maybe that is wrong?
`
FluentValidationModelValidatorProvider
				.Configure(config, provider => provider.ValidatorFactory = container.Resolve<IValidatorFactory>());`

**Comment by JeremySkinner** [comment]
I don't think that's the problem. If that was the issue then your UnityValidatorFactory's GetValidator method wouldn't be called at all. If you could put together that sample, then I'll take a look at it.

**Comment by davidmrichter** [comment]
But why does it use the FluentValidationBodyModelValidator in any case? This one never cares about custom validators.

`public static void Configure(HttpConfiguration configuration, Action<FluentValidationModelValidatorProvider> configurationExpression = null) {
			configurationExpression = configurationExpression ?? delegate { };

			var provider = new FluentValidationModelValidatorProvider();
			configurationExpression(provider);
		    configuration.Services.Replace(typeof(IBodyModelValidator), new FluentValidationBodyModelValidator());
			configuration.Services.Add(typeof(ModelValidatorProvider), provider);
		}`

**Comment by JeremySkinner** [comment]
> But why does it use the FluentValidationBodyModelValidator in any case

It's to work around a bug in WebApi to do with generating error messages - it's not relevant to your issue.  The default BodyModelValidator doesn't play nicely with FluentValidation. The FluentValidationBodyModelValidator has a special case for running FluentValidation's validators that make's sure that property names are built correctly. 

> This one never cares about custom validators.
Instantiation of the validators isn't handled by the BodyModelValidator - it's handled by the provider.

**Comment by davidmrichter** [comment]
Is there any Code Example with WebApi, OWin, custom validator and some IOC Container?

**Comment by JeremySkinner** [comment]
No, sorry

**Comment by SeanKilleen** [comment]
Cleanup: Closing this as it hasn't been active in some time. Feel free to re-open if necessary.
